collect/central_collector.go

```diff
@@ -151,11 +151,74 @@ func (c *CentralCollector) Start() error {
 	return nil
 }
 
+// Stop will be called when the refinery is shutting down.
 func (c *CentralCollector) Stop() error {
 	close(c.done)
 	close(c.incoming)
 	close(c.reload)
-	return c.eg.Wait()
+
+	if err := c.eg.Wait(); err != nil {
+		c.Logger.Error().Logf("error waiting for goroutines to finish: %s", err)
+	}
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, c.Config.GetCollectionConfig().GetShutdownDelay())
+	defer cancel()
+
+	if err := c.shutdown(ctx); err != nil {
+		c.Logger.Error().Logf("error shutting down collector: %s", err)
+	}
+
+	return nil
+}
+
+// shutdown implements the shutdown logic for the collector.
+// It will immediately stop the receiver and the decider.
+// The processor will finish its current cycle before stopping.
+// It will also upload all the remaining traces in the cache to
+// the central store.
+func (c *CentralCollector) shutdown(ctx context.Context) error {
+	// try to send the remaining traces to the cache
+	interval := 20 * time.Microsecond
```

Comment:
This feels like a *really* short interval for an actual network. 

Also, I'm slightly confused here; the comment implies that the processor shuts down after its current run, but it looks like you're allowing the processor to try to send everything. Am I misreading this? Maybe some more comments?

Reply:
Oops, that's my bad. I hardcoded it to make it faster for testing purpose. My original thought was to give a separate `done` signal for the processor so we don't have to spin up a new one. I changed my mind since that would require us adding code to synchronize config change for the interval. I will update the comment to reflect the current design